modules/platforms/cpp/client/src/node_connection.h

```diff
@@ -0,0 +1,174 @@
+#include <atomic>
+#include <mutex>
+#include <unordered_map>
+
+#include "common/utils.h"
+
+#include "ignite/protocol/reader.h"
+#include "ignite/protocol/writer.h"
+#include "ignite/network/async_client_pool.h"
+#include "ignite/ignite_client_configuration.h"
+
+#include "client_operation.h"
+#include "response_handler.h"
+
+namespace ignite::detail
+{
+
+class ClusterConnection;
+
+/**
+ * Represents connection to the cluster.
+ *
+ * Considered established while there is connection to at least one server.
+ */
+class NodeConnection
+{
+    friend class ClusterConnection;
+public:
+    // Deleted
+    NodeConnection() = delete;
+    NodeConnection(NodeConnection&&) = delete;
+    NodeConnection(const NodeConnection&) = delete;
+    NodeConnection& operator=(NodeConnection&&) = delete;
+    NodeConnection& operator=(const NodeConnection&) = delete;
+
+    /**
+     * Destructor.
+     */
+    ~NodeConnection();
+
+    /**
+     * Constructor.
+     *
+     * @param id Connection ID.
+     * @param pool Connection pool.
+     * @param logger Logger.
+     */
+    NodeConnection(uint64_t id, std::shared_ptr<network::AsyncClientPool> pool, std::shared_ptr<IgniteLogger> logger);
+
+    /**
+     * Get connection ID.
+     *
+     * @return ID.
+     */
+    [[nodiscard]]
+    uint64_t getId() const
+    {
+        return m_id;
+    }
+
+    /**
+     * Send request.
+     *
+     * @tparam T Result type.
+     * @param op Operation code.
+     * @param wr Writer function.
+     * @param rd Reader function.
+     * @return Future result.
+     */
+    template<typename T>
+    bool performRequest(ClientOperation op, const std::function<void(protocol::Writer&)>& wr,
+        std::shared_ptr<ResponseHandlerImpl<T>> handler)
+    {
+        auto reqId = generateRequestId();
+        std::vector<std::byte> message;
+        {
+            protocol::BufferAdapter buffer(message);
+            buffer.reserveLengthHeader();
+
+            protocol::Writer writer(buffer);
+            writer.write(int32_t(op));
+            writer.write(reqId);
+            wr(writer);
+
+            buffer.writeLengthHeader();
+
+            {
+                std::lock_guard<std::mutex> lock(m_requestHandlersMutex);
+                m_requestHandlers[reqId] = std::move(handler);
+            }
+        }
+
+        bool sent = m_pool->send(m_id, std::move(message));
+        if (!sent)
+        {
+            std::lock_guard<std::mutex> lock(m_requestHandlersMutex);
+            getAndRemoveHandler(reqId);
+
+            return false;
+        }
+        return true;
+    }
+
+private:
+    /**
+     * Callback that called when new message is received.
+     *
+     * @param msg Received message.
+     */
+    void processMessage(BytesView msg);
+
+    /**
+     * Generate next request ID.
+     *
+     * @return New request ID.
+     */
+    [[nodiscard]]
+    int64_t generateRequestId()
+    {
+        return m_reqIdGen.fetch_add(1, std::memory_order_seq_cst);
```

Comment:
`std::memory_order_relaxed` should be fine here.

Reply:
Agree, just used the default and then forgot to relax it.